server/block/frosted_ice.go

```diff
@@ -0,0 +1,131 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/item/enchantment"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+	"time"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// FrostedIce is a translucent solid block. It is generated by players walking
+// near water while wearing boots enchanted with frost walker.
+type FrostedIce struct {
+	solid
+	
+	// Age is the melting state of frosted ice. Values range from 0 to 4.
+	// When the age is 4 however, the block melts completely into water.
+	Age int
+}
+
+// BreakInfo ...
+func (fi FrostedIce) BreakInfo() BreakInfo {
+	effective := func(_ item.Tool) bool { return true }
+
+	drops := func(_ item.Tool, _ []item.Enchantment) []item.Stack { return []item.Stack{} }
+
+	bi := newBreakInfo(0.5, alwaysHarvestable, effective, drops)
+
+	breakHandler := func(pos cube.Pos, w *world.World, u item.User) {
+		mainHandItem, _ := u.HeldItems()
+
+		if _, isSilkTouch := mainHandItem.Enchantment(enchantment.SilkTouch{}); isSilkTouch {
+			return
+		}
+
+		w.SetBlock(pos, Water{ Still: true, Depth: 8 }, nil)
+	}
+
+	bi.BreakHandler = breakHandler
+	return bi
+}
+
+// Friction ...
+func (fi FrostedIce) Friction() float64 {
+	return 0.98
+}
+
+// LightDiffusionLevel ...
+func LightDiffusionLevel() uint8 {
+	return 3
+}
+
+// Tick ...
+func (fi FrostedIce) Tick(_ int64, pos cube.Pos, w *world.World) {
+	// Tick block once every one and a half seconds.
+	if rand.Intn(19) != 0 {
+		return
+	}
+
+	// One of two conditions must be met in order to increase the age of frosted ice.
```

Comment:
Both your code & the wiki suggest that both conditions have to be met.

Reply:
"If a 1⁄3 chance succeeds or the block has fewer than 4 frosted ice blocks in the 6 blocks surrounding it, and the light level (including light from the sun) at the block would be greater than 11 minus the age, the age is incremented."

I have interpreted this as:  (1⁄3 chance || less than 4) && light levels

The comma after 'surrounding it' would suggest this, no? Furthermore, I have tested frost walker in vanilla bedrock, and at night time the frosted ice does not age at all, suggesting that my interpretation is correct as aging will not occur unless the light levels are above a certain number. My code inverts the statement I wrote above.

I did however notice something quite unusual when testing on the dragonfly server. If I place a frosted ice block above the ground (so that it is not surrounded by any other blocks) it begins melting as it usually would. The same happens if I place a bucket of water above ground and then jump to freeze the source, it begins melting, even at night, when it shouldn't. I suspect there is something going on with the way dragonfly calculates light.

Reply:
My comment describing the conditions is however incorrect, I mended it.